hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/TxnReservedCapacityCounterImpl.java

```diff
@@ -0,0 +1,144 @@
+        this.nodeWideUsedCapacityCounter = nodeWideUsedCapacityCounter;
+        this.reservedCapacityCountByTxId = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Increments 2 counters for supplied txnId.
+     * One is record-store's reserved capacity counter and other
+     * one is node-wide capacity counter. When node-wide
+     * capacity counter is exceeded preconfigured node-wide
+     * limit this method throws {@link ReachedMaxSizeException}
+     *
+     * @throws ReachedMaxSizeException (only when backup is false)
+     */
+    @Override
+    public void increment(UUID txnId, boolean backup) {
+        reservedCapacityCountByTxId.compute(txnId, (ignored, currentCapacityCount) -> {
+            if (backup) {
+                nodeWideUsedCapacityCounter.add(1L);
+            } else {
+                nodeWideUsedCapacityCounter.addCapacityOrThrowException(1);
+            }
+            return currentCapacityCount == null ? 1L : (currentCapacityCount + 1L);
+        });
+    }
+
+    /**
+     * Increments 2 counters with reserved capacities per
+     * txnId. One is record-store's reserved capacity
+     * counter and other one is node-wide capacity counter.
+     *
+     * Note that this method is only used with migrations
+     * and it doesn't throw ReachedMaxSizeException.
+     *
+     * @param reservedCapacityPerTxnId reserved capacities per
+     *                                 txnId
+     */
+    @Override
+    public void copy(Map<UUID, Long> reservedCapacityPerTxnId) {
+        if (isNullOrEmpty(reservedCapacityPerTxnId)) {
+            return;
+        }
+
+        for (Long count : reservedCapacityPerTxnId.values()) {
+            nodeWideUsedCapacityCounter.add(count);
+        }
+        reservedCapacityCountByTxId.putAll(reservedCapacityPerTxnId);
+    }
+
+    /**
+     * Decrements 2 counters for supplied txnId.
+     * One is record-store's reserved capacity counter
+     * and other one is node-wide capacity counter.
+     *
+     * @param txnId id of transaction
+     */
+    @Override
+    public void decrement(UUID txnId) {
+        decrement0(txnId, true);
+    }
+
+    private void decrement0(UUID txnId, boolean decrementNodeWideCounter) {
+        reservedCapacityCountByTxId.computeIfPresent(txnId, (ignored, currentCapacityCount) -> {
+            if (decrementNodeWideCounter) {
+                nodeWideUsedCapacityCounter.add(-1L);
+            }
+            return currentCapacityCount == 1L ? null : currentCapacityCount - 1L;
+        });
+    }
+
+    @Override
+    public void decrementOnlyReserved(UUID txnId) {
+        decrement0(txnId, false);
+    }
+
+    @Override
+    public boolean hasReservedCapacity(UUID txnId) {
+        if (txnId == null) {
+            return false;
+        }
+
+        return reservedCapacityCountByTxId.containsKey(txnId);
+    }
+
+    @Override
+    public void release() {
```

Comment:
Maybe `releaseReservation` would be a better name?

Reply:
renamed to `releaseAllReservations`.